internal/mqini/mqini.go

```diff
@@ -76,3 +88,168 @@ func GetErrorLogDirectory(qm *QueueManager) string {
+// qmgr's ini file path
+func GetIniFilePath(inifilename string, qm *QueueManager) string {
+	var inipath string
+
+    if strings.HasSuffix(inifilename, "qm.ini") {
+		return filepath.Join(qm.Prefix, "qmgrs", qm.Directory, "qm.ini")
+    } else if strings.HasSuffix(inifilename, "mqs.ini") {
+		return filepath.Join(qm.Prefix,"/mqs.ini")
+    } else if strings.HasSuffix(inifilename, "mqat.ini") {
+		return filepath.Join(qm.Prefix,"/mqat.ini")
+	} 
+	return inipath
+}
+
+func SetQMConfigStr(config string) {
+	qmconfigStr = config
+}
+func GetQMConfigStr() string {
+	return qmconfigStr
+}
+
+func WriteToIniFile(userconfig string, inifilepath string) error {
+
+	stanzaList := make(map[string]strings.Builder)
+	var sbAppend strings.Builder
+	var sbMerger strings.Builder
+
+    //No ini file supplied, so nothing to do.
+	if len(userconfig) == 0 {
+		return errors.New("User config supplied was empty")
+	}
+
+    scanner := bufio.NewScanner(strings.NewReader(userconfig))
+	scanner.Split(bufio.ScanLines)
+    consumetoAppend := false
+    consumeToMerge := false
+	var stanza string
+
+    //read through the user file and prepare what we want.
+	for scanner.Scan() {
+		if (strings.Contains(scanner.Text(), ":")) {
+            consumetoAppend=false
+			consumeToMerge=false			
+			stanza = scanner.Text()
+			
+            //check if this stanza exists in the qm.ini
+            if strings.Contains(qmconfigStr,stanza){
+				consumeToMerge=true
+				sbMerger = strings.Builder{}
+				stanzaList[stanza]= sbMerger
+			} else {
+                sbAppend.WriteString(stanza+"\n")
+                consumetoAppend=true
+            }
+        } else {
+            if consumetoAppend {
+                sbAppend.WriteString(scanner.Text()+"\n")
+            }
+            if consumeToMerge {
+				sb := stanzaList[stanza]
+				sb.WriteString(scanner.Text()+"\n")
+				stanzaList[stanza]=sb
+            }
+        }
+    }
+
+	//merge if stanza exits.
+	if len(stanzaList) > 0 {
+		for key, _ := range stanzaList {
+			attrList := stanzaList[key]
+			lineScanner := bufio.NewScanner(strings.NewReader(attrList.String()))
+			lineScanner.Split(bufio.ScanLines)
+			for lineScanner.Scan() {
+		                                                   		attrLine := lineScanner.Text()
+				keyvalue := strings.Split(attrLine,"=")
+				//this line present in qm.ini, update value.
+				if strings.Contains(qmconfigStr, keyvalue[0]) {
+					re := regexp.MustCompile(keyvalue[0]+"=.*")
+					qmconfigStr = re.ReplaceAllString(qmconfigStr, attrLine)
+				} else { //this line not present in qm.ini file, add it.
+					re := regexp.MustCompile(key)
+					newVal := key+"\n"+attrLine
+					qmconfigStr = re.ReplaceAllString(qmconfigStr, newVal)
+				}
+			}
+		}
+	}
+
+	//append if stanza doesn't exist.
+	if len(sbAppend.String()) > 0 {
+       qmconfigStr = qmconfigStr + sbAppend.String()
+	}
+
+	//If this is a unit-test call, we don't write, just return.
+	if isUnitTest {
```

Comment:
Having the production code have a flag for `isUnitTest` doesn't seem right to me.  Couldn't we just refactor the functions a bit, to separate the writing into a separate function?  Then we could unit test all the INI file manipulation as pure Strings, and the prod code could call a wrapper function which does the string manipulation, then writes it to a file.

Reply:
Thanks, I have now removed the flag in my latest commit.